src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py

```diff
@@ -434,55 +438,98 @@ def _resolve_service_principal(client, identifier):
-                                         available_to_other_tenants=False,
-                                         password=password,
-                                         start_date=start_date,
-                                         end_date=end_date)
-    #pylint: disable=no-member
-    aad_sp = _create_service_principal(aad_application.app_id, bool(scopes))
-    oid = aad_sp.output.object_id if scopes else aad_sp.object_id
-
-    if scopes:
-        #It is possible the SP has not been propagated to all servers, so creating assignments
-        #might fail. The reliable workaround is to call out the server where creation occurred.
-        #pylint: disable=protected-access
+    import time
+    graph_client = _graph_client_factory()
+    role_client = _auth_client_factory().role_assignments
+    role = role or 'contributor'
+    scopes = scopes or ['/subscriptions/' + role_client.config.subscription_id]
+    session_key = None
+    sp_oid = None
+    sp_created = False
+    if name:
+        query_exp = 'servicePrincipalNames/any(x:x eq \'{}\')'.format(name)
+        aad_sps = list(graph_client.service_principals.list(filter=query_exp))
+        if aad_sps:
+            sp_oid = aad_sps[0].object_id
+            app_id = aad_sps[0].app_id
+
+    #pylint: disable=protected-access
+    if not sp_oid:
+        start_date = datetime.datetime.utcnow()
+        app_display_name = 'azure-cli-' + start_date.strftime('%Y-%m-%d-%H-%M-%S')
+        if name is None:
+            name = 'http://' + app_display_name # just a valid uri, no need to exist
+
+        end_date = start_date + relativedelta(years=years)
+        password = password or str(uuid.uuid4())
+        aad_application = create_application(graph_client.applications,
+                                             display_name=app_display_name, #pylint: disable=too-many-function-args
+                                             homepage='http://'+app_display_name,
+                                             identifier_uris=[name],
+                                             available_to_other_tenants=False,
+                                             password=password,
+                                             start_date=start_date,
+                                             end_date=end_date)
+        #pylint: disable=no-member
+        app_id = aad_application.app_id
+        #retry while the root cause is being investigated by AAD service team
+        #And the 'session_key' will not be used as well.
+        for _ in range(1, 12):
+            try:
+                aad_sp = _create_service_principal(app_id, bool(scopes), resolve_app=False)
+                break
+            except Exception as ex: #pylint: disable=broad-except
+                #pylint: disable=line-too-long
+                if 'The appId of the service principal does not reference a valid application object' in str(ex):
+                    time.sleep(5)
```

Comment:
Consider: Add logger.warning message indicating that it is retrying something and what retry number you're currently on e.g. (Retry 3/12).